File: pwa/src/lib/grocery/aisleMapper.ts
export type AisleSection = 'Vegetables' | 'Meat' | 'Dairy' | 'Bakery' | 'Pantry';

const AISLE_KEYWORDS: Record<AisleSection, string[]> = {
  Vegetables: [
    'lettuce',
    'spinach',
    'kale',
    'tomato',
    'carrot',
    'broccoli',
    'onion',
    'garlic',
    'celery',
    'cucumber',
    'bell pepper',
    'zucchini',
    'squash',
    'radish',
    'parsnip',
    'potato',
    'sweet potato',
    'cabbage',
    'cauliflower',
    'leek',
    'chard',
    'arugula',
    'bean',
    'pea',
    'eggplant',
    'artichoke',
    'okra',
    'asparagus',
    'corn',
    'green bean',
  ],
  Meat: [
    'beef',
    'ground beef',
    'steak',
    'chuck',
    'brisket',
    'chicken',
    'breast',
    'thigh',
    'leg',
    'wing',
    'pork',
    'ham',
    'bacon',
    'sausage',
    'chop',
    'lamb',
    'mutton',
    'fish',
    'salmon',
    'cod',
    'tuna',
    'shrimp',
    'prawn',
    'turkey',
    'duck',
    'goose',
  ],
  Dairy: [
    'milk',
    'cream',
    'butter',
    'cheese',
    'yogurt',
    'sour cream',
    'cottage cheese',
    'mozzarella',
    'cheddar',
    'parmesan',
    'feta',
    'ricotta',
    'brie',
    'gouda',
    'egg',
    'eggs',
    'cream cheese',
  ],
  Bakery: [
    'bread',
    'flour',
    'baguette',
    'croissant',
    'pastry',
    'bagel',
    'roll',
    'bun',
    'tortilla',
    'pita',
    'naan',
    'focaccia',
    'sourdough',
    'rye',
  ],
  Pantry: [
    'oil',
    'olive oil',
    'salt',
    'pepper',
    'sugar',
    'spice',
    'paprika',
    'cumin',
    'cinnamon',
    'oregano',
    'basil',
    'thyme',
    'rosemary',
    'dill',
    'parsley',
    'rice',
    'pasta',
    'beans',
    'lentil',
    'chickpea',
    'canned',
    'sauce',
    'broth',
    'vinegar',
    'soy sauce',
    'sesame oil',
    'coconut oil',
    'baking',
    'powder',
    'soda',
  ],
};

function stringSimilarity(a: string, b: string): number {
  const aLower = a.toLowerCase();
  const bLower = b.toLowerCase();

  if (aLower === bLower) return 1;
  if (aLower.includes(bLower) || bLower.includes(aLower)) return 0.8;

  let matches = 0;
  for (const char of aLower) {
    if (bLower.includes(char)) matches++;
  }
  return matches / Math.max(aLower.length, bLower.length);
}

export function mapIngredientToAisle(ingredientName: string): AisleSection {
  const lowerIngredient = ingredientName.toLowerCase();

  // First, try exact or near-exact matches with keywords
  for (const [aisle, keywords] of Object.entries(AISLE_KEYWORDS)) {
    for (const keyword of keywords) {
      if (lowerIngredient.includes(keyword)) {
        return aisle as AisleSection;
      }
    }
  }

  // Fallback: use fuzzy matching to find best aisle
  let bestAisle: AisleSection = 'Pantry';
  let bestScore = 0;

  for (const [aisle, keywords] of Object.entries(AISLE_KEYWORDS)) {
    for (const keyword of keywords) {
      const score = stringSimilarity(lowerIngredient, keyword);
      if (score > bestScore && score > 0.6) {
        bestScore = score;
        bestAisle = aisle as AisleSection;
      }
    }
  }

  return bestAisle;
}

export function groupIngredientsByAisle(ingredients: string[]): Record<AisleSection, string[]> {
  const aisles: Record<AisleSection, string[]> = {
    Vegetables: [],
    Meat: [],
    Dairy: [],
    Bakery: [],
    Pantry: [],
  };

  for (const ingredient of ingredients) {
    const aisle = mapIngredientToAisle(ingredient);
    aisles[aisle].push(ingredient);
  }

  return aisles;
}

File: api/src/RecipeApi/Services/AisleMapper.cs
using System;
using System.Collections.Generic;

namespace RecipeApi.Services;

public enum GrocerySection
{
    Produce,
    Meat,
    Seafood,
    DairyAndEggs,
    Frozen,
    Bakery,
    Pantry,
    Beverages,
    Deli,
    Grocery
}

/// <summary>
/// Maps a canonical ingredient name to a <see cref="GrocerySection"/> using
/// longest-match keyword lookup. Falls back to <see cref="GrocerySection.Grocery"/>
/// when no keyword matches.
/// </summary>
public class AisleMapper
{
    // Section priority order used to break ties when two keywords produce the same match length.
    private static readonly GrocerySection[] SectionPriority =
    [
        GrocerySection.Produce,
        GrocerySection.Meat,
        GrocerySection.Seafood,
        GrocerySection.DairyAndEggs,
        GrocerySection.Frozen,
        GrocerySection.Bakery,
        GrocerySection.Beverages,
        GrocerySection.Deli,
        GrocerySection.Pantry,
    ];

    // Keyword sets per section (English + French).
    // Broth/stock are intentionally in Beverages, not Pantry.
    private static readonly Dictionary<GrocerySection, string[]> Keywords = new()
    {
        [GrocerySection.Produce] =
        [
            // English
            "bell pepper", "sweet potato", "blueberry", "strawberry", "cauliflower",
            "asparagus", "cucumber", "zucchini", "broccoli", "mushroom", "avocado",
            "jalapeño", "rosemary", "cilantro", "parsley", "cabbage", "spinach",
            "lettuce", "tomato", "carrot", "potato", "banana", "orange", "celery",
            "onion", "garlic", "lemon", "mango", "berry", "grape", "peach", "thyme",
            "basil", "chive", "apple", "leek", "corn", "lime", "kale", "pear",
            "herb", "dill", "mint", "ginger", "pepper",
            // French
            "poivron", "patate douce", "myrtille", "fraise", "chou-fleur",
            "asperge", "concombre", "courgette", "brocoli", "champignon", "avocat",
            "jalapeño", "romarin", "coriandre", "persil", "chou", "épinard",
            "laitue", "tomate", "carotte", "pomme de terre", "banane", "orange",
            "céleri", "oignon", "ail", "citron", "mangue", "baie", "raisin",
            "pêche", "thym", "basilic", "ciboulette", "pomme", "poireau", "maïs",
            "citron vert", "chou frisé", "poire", "herbe", "aneth", "menthe",
            "gingembre", "poivre",
        ],

        [GrocerySection.Meat] =
        [
            // English
            "ground beef",
            "beef", "steak", "chuck", "brisket", "chicken", "breast", "thigh",
            "wing", "pork", "ham", "bacon", "sausage", "lamb", "mutton",
            "turkey", "duck", "veal", "venison", "leg",
            // French
            "boeuf haché",
            "boeuf", "steak", "paleron", "poitrine", "poulet", "blanc", "cuisse",
            "aile", "porc", "jambon", "lardons", "saucisse", "agneau", "mouton",
            "dinde", "canard", "veau", "chevreuil",
        ],

        [GrocerySection.Seafood] =
        [
            // English
            "smoked salmon",
            "fish", "salmon", "cod", "tuna", "tilapia", "halibut", "shrimp",
            "prawn", "scallop", "crab", "lobster",
            // French
            "saumon fumé",
            "poisson", "saumon", "cabillaud", "thon", "tilapia", "flétan",
            "crevette", "pétoncle", "crabe", "homard",
        ],

        [GrocerySection.DairyAndEggs] =
        [
            // English
            "cream cheese", "cottage cheese", "sour cream", "half-and-half",
            "whipping cream", "mozzarella", "parmesan", "cheddar", "ricotta",
            "gouda", "feta", "brie", "kefir", "yogurt", "butter", "cheese",
            "cream", "milk", "egg",
            // French
            "fromage à la crème", "fromage cottage", "crème sure", "demi-crème",
            "crème fouettée", "mozzarella", "parmesan", "cheddar", "ricotta",
            "gouda", "feta", "brie", "kéfir", "yaourt", "beurre", "fromage",
            "crème", "lait", "oeuf",
        ],

        [GrocerySection.Frozen] =
        [
            // English
            "frozen peas", "frozen corn", "frozen berries", "frozen pizza",
            "ice cream", "gelato", "sorbet", "edamame", "frozen",
            // French
            "petits pois surgelés", "maïs surgelé", "baies surgelées",
            "pizza surgelée", "crème glacée", "gelato", "sorbet", "edamame",
            "surgelé", "surgele",
        ],

        [GrocerySection.Bakery] =
        [
            // English
            "rye bread", "sourdough", "focaccia", "baguette", "croissant",
            "brioche", "tortilla", "pastry", "bagel", "bread", "scone",
            "muffin", "naan", "pita", "roll", "bun",
            // French
            "pain de seigle", "pain au levain", "focaccia", "baguette",
            "croissant", "brioche", "tortilla", "viennoiserie", "bagel",
            "pain", "scone", "muffin", "naan", "pita",
        ],

        [GrocerySection.Beverages] =
        [
            // English — broth/stock intentionally here, not Pantry
            "sparkling water", "coconut milk", "almond milk", "oat milk",
            "broth", "stock", "juice", "cider", "coffee", "wine", "beer",
            "soda", "tea",
            // French
            "eau gazeuse", "lait de coco", "lait d'amande", "lait d'avoine",
            "bouillon", "fond", "jus", "cidre", "café", "vin", "bière",
            "soda", "thé",
        ],

        [GrocerySection.Deli] =
        [
            // English
            "smoked salmon", "prepared salad", "prosciutto", "mortadella",
            "pastrami", "pepperoni", "rotisserie", "salami", "hummus", "deli",
            // French
            "saumon fumé", "salade préparée", "prosciutto", "mortadelle",
            "pastrami", "pepperoni", "poulet rôti", "salami", "houmous",
            "charcuterie",
        ],

        [GrocerySection.Pantry] =
        [
            // English
            "vinegar", "baking", "pasta", "sauce", "flour", "sugar", "spice",
            "rice", "salt", "oil",
            // French
            "vinaigre", "levure", "pâtes", "sauce", "farine", "sucre",
            "épice", "riz", "sel", "huile",
        ],
    };

    /// <summary>
    /// Maps a canonical ingredient name to a <see cref="GrocerySection"/> using
    /// longest-match keyword lookup. Returns <see cref="GrocerySection.Grocery"/>
    /// when no keyword matches.
    /// </summary>
    public GrocerySection MapToSection(string canonicalName)
    {
        if (string.IsNullOrWhiteSpace(canonicalName))
            return GrocerySection.Grocery;

        var normalized = canonicalName.ToLowerInvariant();

        int bestLength = 0;
        GrocerySection bestSection = GrocerySection.Grocery;

        // Iterate in priority order so ties are broken by section priority.
        foreach (var section in SectionPriority)
        {
            if (!Keywords.TryGetValue(section, out var keywords))
                continue;

            foreach (var keyword in keywords)
            {
                if (normalized.Contains(keyword, StringComparison.OrdinalIgnoreCase))
                {
                    if (keyword.Length > bestLength)
                    {
                        bestLength = keyword.Length;
                        bestSection = section;
                    }
                }
            }
        }

        return bestSection;
    }
}
